Sources/NIOHTTP1/HTTPTypedPipelineSetup.swift

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+@_spi(AsyncChannel) import NIOCore
+
+// MARK: - Server pipeline configuration
+
+/// Configuration for an upgradable HTTP pipeline.
+@available(macOS 13, iOS 16, tvOS 16, watchOS 9, *)
+@_spi(AsyncChannel)
+public struct NIOUpgradableHTTPServerPipelineConfiguration<UpgradeResult: Sendable> {
+    /// Whether to provide assistance handling HTTP clients that pipeline
+    /// their requests. Defaults to `true`. If `false`, users will need to handle clients that pipeline themselves.
+    public var enablePipelining = true
+
+    /// Whether to provide assistance handling protocol errors (e.g. failure to parse the HTTP
+    /// request) by sending 400 errors. Defaults to `true`.
+    public var enableErrorHandling = true
+
+    /// Whether to validate outbound response headers to confirm that they are
+    /// spec compliant. Defaults to `true`.
+    public var enableResponseHeaderValidation = true
+
+    /// The configuration for the ``HTTPResponseEncoder``.
+    public var httpResponseEncoderConfiguration = HTTPResponseEncoder.Configuration()
+
+    /// The configuration for the ``NIOTypedHTTPServerUpgradeHandler``.
+    public var httpServerUpgradeConfiguration: NIOTypedHTTPServerUpgradeConfiguration<UpgradeResult>
```

Comment:
do we need the `http` prefixes here? I think the context from type name should be enough. Can probably drop the `Server` from the `httpServerUpgradeConfiguration` as well.